src/main/java/hudson/tasks/junit/pipeline/JUnitResultsStepExecution.java

```diff
@@ -0,0 +1,83 @@
+import hudson.tasks.junit.TestResultAction;
+import hudson.tasks.junit.TestResultSummary;
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.actions.ThreadNameAction;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.Filterator;
+import org.jenkinsci.plugins.workflow.graphanalysis.FlowScanningUtils;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution;
+import org.jenkinsci.plugins.workflow.support.steps.StageStep;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JUnitResultsStepExecution extends SynchronousNonBlockingStepExecution<TestResultSummary> {
+
+    private transient final JUnitResultsStep step;
+
+    public JUnitResultsStepExecution(@Nonnull JUnitResultsStep step, StepContext context) {
+        super(context);
+        this.step = step;
+    }
+
+    @Override
+    protected TestResultSummary run() throws Exception {
+        FilePath workspace = getContext().get(FilePath.class);
+        workspace.mkdirs();
+        Run<?,?> run = getContext().get(Run.class);
+        TaskListener listener = getContext().get(TaskListener.class);
+        Launcher launcher = getContext().get(Launcher.class);
+        FlowNode node = getContext().get(FlowNode.class);
+
+        String nodeId = node.getId();
+
+        TestResultAction testResultAction = JUnitResultArchiver.parseAndAttach(step, nodeId, getEnclosingStagesAndParallels(node),
+                run, workspace, launcher, listener);
+
+        if (testResultAction != null) {
+            // TODO: Once JENKINS-43995 lands, update this to set the step status instead of the entire build.
+            if (testResultAction.getResult().getFailCount() > 0) {
+                getContext().setResult(Result.UNSTABLE);
+            }
+            return new TestResultSummary(testResultAction.getResult().getResultByRunAndNode(run.getExternalizableId(), nodeId));
+        }
+
+        return new TestResultSummary();
+    }
+
+    /**
+     * Get the stage and parallel branch start node IDs (not the body nodes) for this node, innermost first.
+     * @param node A flownode.
+     * @return A nonnull, possibly empty list of stage/parallel branch start node IDs, innermost first.
+     */
+    @Nonnull
+    public static List<String> getEnclosingStagesAndParallels(FlowNode node) {
+        List<String> enclosingBlocks = new ArrayList<>();
+        for (FlowNode enclosing : node.getEnclosingBlocks()) {
+            if (enclosing != null && enclosing.getAction(LabelAction.class) != null) {
+                if ((enclosing instanceof StepStartNode && ((StepStartNode) enclosing).getDescriptor() instanceof StageStep.DescriptorImpl) ||
```

Comment:
will this not blow up if `pipeline-stage-step` is not installed (it is optional)...

Reply:
Yes. Either we make the pipeline dependencies optional and accept that things blow up (`FlowNode` from `workflow-api` too!) or we make them non optional.

Reply:
Move the specific pipeline step executor to a separate plugin like
* docker-build-step-plugin
* workflow-step-api-plugin
* pipeline-maven-plugin
* XYZ-build-step-api-plugin
and so on

Reply:
...why?

Reply:
for the same reason those plugin are created. Move pipeline dependency far from base plugin. For the same reason do not force dependency from Jenkins 2.7.3 because
> So the main reason for doing so is that there'll be no real benefit for users on <2.7.3 - since the reporting will be in Blue Ocean.
than move this step to one of Blue Ocean plugin.

Reply:
Those plugins don’t duplicate existing behavior, though. This is the right place for a junit step.

Reply:
A similar concept was expressed by jglick [here](https://github.com/jenkinsci/junit-plugin/pull/9#issuecomment-68792525)

Reply:
That was 2.5 years ago - a lot has changed since then, most notably the end of bundling.

Reply:
@abayer so this should not blow up.  If you can not use this step without the `pipeline-stage-step` then the step should not be active (ie you can not use it), or you can use it but you get no "per stage" results.

You should be able to do this, there are many ways to do it.
one would be to check if the plugin is installed and only then call some code in a different class.
I'm not sure if Steps can be marked as Optional using the varient plugin or the like to say this is only active if this plugin is installed, but I guess it should be possible, if not I think we should be able to do this.

Reply:
Going with the `variant` approach.

Reply:
I don't understand the variant approach but if it deals with this I'm all for it.

Reply:
Actually @abayer  you *could* just have it look for the descriptor *ID* matching, which means you don't need a hard dependency on stage-step (string equals vs instanceof).